### PR TITLE
add karmada operator apis

### PR DESCRIPTION
**Pull request by calvin0327**
Signed-off-by: calvin0327 <wen.chen@daocloud.io>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add karmada operator apis and update scripts to generate template code associated with apis.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
the all go files associated with operator is locate to the dir `operator`. 
It ref to  https://github.com/istio/istio/tree/master/operator

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by calvin0327** [comment]
@lonelyCZ @RainbowMango /cc

**Comment by lonelyCZ** [comment]
Should we generate crd by `hack/update-crdgen.sh`? @calvin0327

**Comment by lonelyCZ** [comment]
Putting them in `operator/config/crd` for the moment?

**Comment by calvin0327** [comment]
> Putting them in `operator/config/crd` for the moment?

Reasonable.

File: hack/update-crdgen.sh [hunk 1]
```diff
@@ -18,3 +18,4 @@ controller-gen crd paths=./pkg/apis/policy/... output:crd:dir=./charts/karmada/_
 controller-gen crd paths=./pkg/apis/work/... output:crd:dir=./charts/karmada/_crds/bases
 controller-gen crd paths=./pkg/apis/networking/... output:crd:dir=./charts/karmada/_crds/bases
 controller-gen crd paths=./examples/customresourceinterpreter/apis/... output:crd:dir=./examples/customresourceinterpreter/apis/
+controller-gen crd paths=./operator/pkg/apis/operator/... output:crd:dir=./operator/config/crds
```
Comment (lonelyCZ) [inline hack/update-crdgen.sh:21]: Please add new line at end.

File: operator/pkg/apis/operator/v1alpha1/type.go [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+	// For supported flags, please see
+	// https://github.com/karmada-io/karmada/blob/master/cmd/agent/app/options/options.go
+	// for details.
+	// +optional
+	ExtraArgs map[string]string `json:"extraArgs,omitempty"`
+}
+
+// KarmadaControlPlane describes how to connect karmada control plane.
+type KarmadaControlPlane struct {
+	// KarmadaName specifield name of karmadaDeployment, it represente
```
Comment (Poor12) [inline operator/pkg/apis/operator/v1alpha1/type.go:736]: KarmadaName specifies the name of karmadaDeployment. It represents

File: operator/pkg/apis/operator/v1alpha1/type.go [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+	// SyncMode describes how a cluster sync resources from karmada control plane.
+	// +kubebuilder:validation:Enum=Push;Pull
+	// +required
+	SyncMode ClusterSyncMode `json:"syncMode"`
+
+	// ClusterName describes the karmada cluster resource name.
+	// +required
+	ClusterName string `json:"clusterName"`
+
+	// ClusterNamespace define namespace in the control plane where member secrets cluster are stored.
```
Comment (Poor12) [inline operator/pkg/apis/operator/v1alpha1/type.go:662]: ClusterNamespace defines the namespace in karmada control plane where secrets of member clusters are stored.

**Review by Poor12 — COMMENTED** [review]
Generally LGTM, some small nits.

**Comment by calvin0327** [comment]
@lonelyCZ @Poor12 
thanks,  updated.

File: operator/pkg/apis/operator/v1alpha1/type.go [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:path=karmadaonborad,scope=Namespaced,categories={karmada-io}
+// +kubebuilder:printcolumn:JSONPath=`.spec.syncMode`,name="Mode",type=string
+// +kubebuilder:printcolumn:JSONPath=`.metadata.creationTimestamp`,name="Age",type=date
+
+// KarmadaOnborad enables declarative join member cluster to karmada.
+type KarmadaOnborad struct {
```
Comment (RainbowMango) [inline operator/pkg/apis/operator/v1alpha1/type.go:622]: Did we ever talk about this API?

Comment (RainbowMango) [reply operator/pkg/apis/operator/v1alpha1/type.go:622]: I know this API is added by https://github.com/karmada-io/karmada-operator/pull/4. 
I think we all missed it from reviewing.

For this PR, I suppose we can drop it and we can add it back when we needed it.

Comment (calvin0327) [reply operator/pkg/apis/operator/v1alpha1/type.go:622]: This API we haven't talked about yet. 
ok, I remove it first

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**